src/relay/ir/Relay.g4

```diff
@@ -0,0 +1,140 @@
+grammar Relay;
+
+// Lexing
+// comments
+WS : [ \t\n\r]+ -> skip ;
+LINE_COMMENT : '//' .*? '\n' -> skip ;
+COMMENT : '/*' .*? '*/' -> skip ;
+
+// operators
+MUL: '*' ;
+DIV: '/' ;
+ADD: '+' ;
+SUB: '-' ;
+LT: '<' ;
+GT: '>' ;
+LE: '<=' ;
+GE: '>=' ;
+EQ: '==' ;
+NE: '!=' ;
+
+operatorIdent: CNAME ;
+GLOBAL_IDENT: '@' CNAME ;
+LOCAL_IDENT: '%' CNAME ;
+
+MUT: 'mut' ;
+
+BOOL_LIT
+  : 'true'
+  | 'false'
+  ;
+
+// non-negative floats
+FLOAT
+  : INT '.' INT EXP? // 1.35, 1.35E-9, 0.3, 4.5
+  | INT EXP // 1e10 3e4
+  ;
+
+// non-negative ints
+// INT: '0' | [1-9] DIGIT* ; // no leading zeros
+INT: DIGIT+ ;
+fragment EXP: [eE] [+\-]? INT ; // \- since - means "range" inside [...]
+
+CNAME: ('_'|LETTER) ('_'|LETTER|DIGIT)* ;
+fragment LETTER: [a-zA-Z] ;
+fragment DIGIT: [0-9] ;
+
+// Parsing
+
+// a program is a list of options followed by either an expression or a list of definitions
+// TODO: is this really the desired outline? if there are defns should there be an expression at the end?
+prog: option* (expr | defn+) EOF ;
+
+option: 'set' ident BOOL_LIT ;
+
+expr
+  // operators
+  // todo: this precedence seems fishy
+  : '(' expr ')'                  # parens
+  | '-' expr                              # neg
+  | expr op=('*'|'/') expr                # binOp
+  | expr op=('+'|'-') expr                # binOp
+  | expr op=('<'|'>'|'<='|'>=') expr      # binOp
+  | expr op=('=='|'!=') expr              # binOp
+
+  // function definition and application
+  | expr '(' (expr (',' expr)*)? ')'      # call
+  | fun                                   # funExpr
+
+  // tuples and tensors
+  | '(' ')'                               # tuple
+  | '(' expr ',' ')'                      # tuple
+  | '(' expr (',' expr)+ ')'              # tuple
+  | '[' (expr (',' expr)*)? ']'           # tensor
+
+  // todo: precedence
+  | 'if' expr body 'else' body            # ifElse
+
+  // sequencingg
+  | 'let' MUT? ident (':' type_)? '=' expr ';' expr  # let
```

Comment:
what does expr ';' expr mean here?

Reply:
Relay doesn't have statements, only expressions. Thus `let x = 1;` is not valid Relay, because it doesn't have a value. `let` expressions must always contain an identity, an assignment, and a body that might use that assignment (which is the value of the let expression). For example,

```
let x = 1;
x
```

is valid Relay and evaluates to `1`.

This is similar to the syntax used by ReasonML, which is a version of OCaml made to look like JavaScript.